hugr-core/src/extension.rs

```diff
@@ -82,13 +86,35 @@ impl ExtensionRegistry {
     /// Returns a reference to the registered extension if successful.
     pub fn register(&mut self, extension: Extension) -> Result<&Extension, ExtensionRegistryError> {
         match self.0.entry(extension.name().clone()) {
-            btree_map::Entry::Occupied(_) => Err(ExtensionRegistryError::AlreadyRegistered(
+            btree_map::Entry::Occupied(prev) => Err(ExtensionRegistryError::AlreadyRegistered(
                 extension.name().clone(),
+                prev.get().version().clone(),
+                extension.version().clone(),
             )),
             btree_map::Entry::Vacant(ve) => Ok(ve.insert(extension)),
         }
     }
 
+    /// Registers a new extension to the registry, keeping most up to date if extension exists.
+    ///
+    /// If extension IDs match, the extension with the higher version is kept.
```

Comment:
I note we are not really doing anything with Semver here (it could be any totally-ordered thing), and I wonder whether this should be the place. Replacing an extension with a breaking-change-upgrade could be problematic; should we
* have a boolean flag here, allow breaking-change upgrades
* keep multiple major versions of the extension (a single minor version of each)? I think this will be too difficult because then we'd have to version every *reference* to an extension.
* Error on any breaking-change upgrade
* Do something at Hugr validation time. In which case maybe just the checks we have are sufficient, so maybe the right answer is to do nothing, i.e. leave PR as it stands....

Reply:
Yes I think I prefer to keep the logic here simple and flexible and deal with valid
updates etc. elsewhere. 

Maybe I should add a "remove" function so the registry is actually fully flexible, and
this is just a utility for update-only changes?

Reply:
added remove